datastore/sqlite.nim

```diff
@@ -136,9 +142,26 @@ template journalModePragmaStmt*(env: SQLite): RawStmtPtr =
     s.dispose
     return failure $sqlite3_errstr(x)
 
-  if (let x = $sqlite3_column_text(s, 0).cstring; x != "memory" and x != "wal"):
+  let
+    x = sqlite3_column_text(s, 0).cstring
+
+  # detect out-of-memory error
+  # see the conversion table and final paragraph of:
+  # https://www.sqlite.org/c3ref/column_blob.html
+  # see also https://www.sqlite.org/rescode.html
+
+  # in order to detect an out-of-memory error check that the result is a null
+  # pointer and that the result code is an error code
+  if x.isNil:
+    let
+      code = sqlite3_errcode(sqlite3_db_handle(s))
+
+    if not (code in [SQLITE_OK, SQLITE_ROW, SQLITE_DONE]):
+      raise (ref Defect)(msg: $sqlite3_errstr(code))
```

Comment:
OOM in Nim runtime itself is unrecoverable. OOM in external program is just a failure. But since SQLite engine is built into executable, I'm not sure.

Probably, it's better to be on the safe side and make it Defect. It probably will be very rare Defect.